File: PeakElement-162.py
# Time Complexity : O(logn)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : One of the rare question of binary search without sorted array


# Your code here along with comments explaining your approach 

class Solution(object):
    def findPeakElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        low = 0
        high = len(nums) - 1
        
        while low <= high:
            mid = low + (high - low)//2
            if ((mid == 0 or nums[mid] > nums[mid-1]) and (mid == len(nums)-1 or nums[mid] > nums[mid+1])):
                return mid
            elif nums[mid+1] > nums[mid]:
                low = mid + 1
            else:
                high = mid - 1
        return -1

s =  Solution()
print s.findPeakElement([1,2,3,1])

File: peakElement162.java
package nov192020;

// Problem 9 
// https://leetcode.com/problems/find-peak-element/
// https://github.com/super30admin/Binary-Search-2
public class peakElement162 {
	public int findPeakElement(int[] nums) {
		int low = 0;
		int high = nums.length - 1;
		while (low < high) {
			int mid = (low + high) / 2;
			if (nums[mid] > nums[mid + 1]) {
				high = mid;
			} else {
				low = mid + 1;
			}
		}
		return low;
	}

}

//. Time Complexity : O(log n)
// Space Complexity : O (1)
